Ensure ssh hostkey checks respect server port

willthames opened the pull request:
##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
module_utils/known_hosts

##### ANSIBLE VERSION
```
ansible 2.3.0 (devel 462ab6b1db) last updated 2017/01/31 10:02:44 (GMT +1000)
  config file = 
  configured module search path = Default w/o overrides
```

Also affects 2.2.1, but not 2.2.0

##### SUMMARY
ssh-keyscan will default to getting the host key for port 22.
If the ssh service is running on a different port, ssh-keyscan
will need to know this.

Tidy up minor flake8 issues

Commit:
Add tests for `get_fqdn_and_port` method.

Currently tests verify original behavior - returning default `ssh-keyscan` port
Add test around `add_host_key` to verify underlying command arguments
Add some new expectations for `get_fqdn_and_port`
Test that non-standard port is passed to `ssh-keyscan` command

Commit:
Ensure ssh hostkey checks respect server port

ssh-keyscan will default to getting the host key for port 22.
If the ssh service is running on a different port, ssh-keyscan
will need to know this.

Tidy up minor flake8 issues

Commit:
Update known_hosts tests for port being None

Ensure that git urls don't try and set port when a path
is specified

Update known_hosts tests to meet flake8

willthames commented:
Updated with @nev3rm0re's tests from #18734 - had to make some changes around ipv6 url parsing to ensure the code was sound

abadger commented:
shippable tests are failing but it looks like unrelated errors.  I've hit rebuild an I'll check back later to see the results.

abadger commented:
Okay, shippable is reporting two errors: 
* The file passes  the current pep8 style checks so it can be removed from the whitelist here: test/sanity/pep8/legacy-files.txt
  * I can fix that once it's merged if needed.
* With this change applied, the lastpass lookup unittests are failing: https://app.shippable.com/runs/5890ada98e4818100037b39d/2/console
  * I have no idea why that's happening.  @mattclay can you take a look?

mattclay commented on test/units/module_utils/basic/test_known_hosts.py:
```diff
@@ -19,37 +19,88 @@
+        args = json.dumps(dict(ANSIBLE_MODULE_ARGS={}))
+        # unittest doesn't have a clean place to use a context manager, so we have to enter/exit manually
+        self.stdin_swap = swap_stdin_and_argv(stdin_data=args)
+        self.stdin_swap.__enter__()
```
This context manager leaks, breaking later tests. You should be able to replace with:

```python
        with swap_stdin_and_argv(stdin_data=args):
```

willthames replied:
Thanks @mattclay - I suspect this was copied from elsewhere, but I'll look at fixing it up today

mattclay replied:
Yes, there are several other tests using `__enter__` instead of `with`. Those tests work because they also call `__exit__`, but that's not ideal.

Commit:
Fix stdin swap context for test_known_hosts

Move test_known_hosts from under basic, as it is its own library.
Remove module_utils.known_hosts from pep8 legacy files list

willthames commented:
@mattclay I was hoping that fixing that would also make `nosetests --pdb` work again (it works if I delete the test with the stdin swap). Don't suppose you have any thoughts on that?

kostaatsk commented:
#20368

mattclay commented:
I merged #21031 which fixes tests which were interfering with the unit tests from this PR. Unit tests are now passing. I've restarted CI to resolve another unrelated test failure.

willthames commented:
@mattclay @abadger are we good to merge this then? And then backport to stable-2.2 too (as 2.2.1 is currently broken with this)

mattclay commented:
If we do backport this, we'll probably need to backport my fix as well, to avoid breaking tests there.

abadger commented:
Okay, merged to devel and cherrypicked to stable-2.2.  mattclay also cherrypicked the test fix back as well.
